### PR TITLE
    libsubprocess: handle remote protocol errors

### PR DESCRIPTION
**Pull request by chu11**
Problem: When handling status changes for remote subprocesses in rexec_continuation(), it is assumed that a process has reached the EXITED state when the rpc stream receives ENODATA.

A recent broker crash shows that this may not always be the case when systemd (i.e. sdexec) is used to launch remote subprocesses.  In this case, the stream ended before the initial PID was reported.  It is assumed under extreme error scenarios, that systemd may not necessarily report information in the order one would expect with a simple fork/exec.

Solution: If the status stream ends early, do not assume the remote process has exited.  Instead, handle this as an EPROTO error and fail the subprocess.

Fixes https://github.com/flux-framework/flux-core/issues/5956

### CODE REVIEW

**Comment by garlick** [comment]
IMHO there are really three issues
1. The fact that the client asserted when the server violated the protocol
2. The fact that sdexec violated the protocol
3. The fact that RFC 42 doesn't document explicitly any required responses before terminating the exec stream with ENODATA (success).

This fixes (1). Instead of calling out the sdexec behavior explicitly as though we are planning to live with it, it should just reference (2).  And (2) can reference RFC 42.

Also, is restoring the assertion really OK?  Is there clearly now no way to hit it?  (I haven't looked at the code in detail)

**Comment by chu11** [comment]
re-pushed, updating comment and commit message per notes above

File: src/common/libsubprocess/remote.c [hunk 1]
```diff
@@ -615,7 +606,21 @@ static void rexec_continuation (flux_future_t *f, void *arg)
...
         if (errno == ENODATA) {
-            remote_completion (p);
+            p->remote_completed = true;
+            /* Per RFC42, when remote processes are launched, the
+             * process should return finished (i.e. state EXITED)
+             * before returning ENODATA.  It is otherwise considered a
+             * protocol error.
+             *
+             * N.B. There is evidence that the sdexec module has
+             * violated the protocol before #6012.
```
Comment (garlick) [inline src/common/libsubprocess/remote.c:616]: I would say the more appropriate ref would be #5956 as that's where the larger issue starts.

File: src/common/libsubprocess/subprocess.c [hunk 2]
```diff
@@ -246,11 +246,7 @@ void subprocess_standard_output (flux_subprocess_t *p, const char *stream)
 
 void subprocess_check_completed (flux_subprocess_t *p)
 {
-    if (p->state != FLUX_SUBPROCESS_EXITED) {
-        log_err ("subprocess_check_completed: unexpected state %s",
-                 flux_subprocess_state_string (p->state));
-        return;
-    }
+    assert (p->state == FLUX_SUBPROCESS_EXITED);
```
Comment (garlick) [inline src/common/libsubprocess/subprocess.c:249]: I think it actually might be wise to not revert this commit.  Based on experience with el cap failures, we don't want to have this assert enabled.  So the question is probably do you want a log message when something weird happens that only happens at scale or do you want nothing at all when @grondo builds the RPM with NDEBUG.

**Review by garlick — APPROVED** [review]
LGTM just some minor suggestions.

**Comment by chu11** [comment]
re-pushed per comments above, although I tweaked the commit message header & text b/c "assert" is no longer true after I chose not to revert the previous change.

**Comment by garlick** [comment]
Looks good, thanks!

**Comment by chu11** [comment]
great, setting MWP

**Pull request merged**